### PR TITLE
introduce FeatureGate to kubescheduleconfig

### LINKED ISSUE #84922: introduce FeatureGates to KubeSchedulerConfiguration
**What would you like to be added**:
introduce FeatureGates to KubeSchedulerConfiguration
/assign
**Why is this needed**:
This threads the --feature-gates flag through the KubeSchedulerConfiguration object

### PR DESCRIPTION
**Pull request by zouyee**
Signed-off-by: Zou Nengren <zouyee1989@gmail.com>

**What type of PR is this?**
 
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #84922

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
introduce FeatureGates to KubeSchedulerConfiguration
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by zouyee** [comment]
@kubernetes/sig-scheduling-api-reviews

**Comment by zouyee** [comment]
@ahg-g PTAL

**Comment by ahg-g** [comment]
/cc @alculquicondor 

I will leave for Aldo to review, who will be looking more closely at Component Config in the context of #85737

File: cmd/kube-scheduler/app/options/options_test.go [hunk 1]
```diff
@@ -628,10 +633,9 @@ pluginConfig:
 				return
 			}
 
-			if !reflect.DeepEqual(config.ComponentConfig, tc.expectedConfig) {
-				t.Errorf("config.diff:\n%s", diff.ObjectReflectDiff(tc.expectedConfig, config.ComponentConfig))
+			if diff := cmp.Diff(config.ComponentConfig, tc.expectedConfig); diff != "" {
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go]: parameters are in the wrong order, according to the error message.

File: cmd/kube-scheduler/app/options/options_test.go [hunk 2]
```diff
@@ -604,6 +608,7 @@ pluginConfig:
 				PodInitialBackoffSeconds: defaultPodInitialBackoffSeconds,
 				PodMaxBackoffSeconds:     defaultPodMaxBackoffSeconds,
 				Plugins:                  nil,
+				FeatureGates:             map[string]bool{},
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go]: can we have a new test case that uses feature gates?

File: pkg/scheduler/apis/config/types.go [hunk 3]
```diff
@@ -110,6 +110,8 @@ type KubeSchedulerConfiguration struct {
 	// PluginConfig is an optional set of custom plugin arguments for each plugin.
 	// Omitting config args for a plugin is equivalent to using the default config for that plugin.
 	PluginConfig []PluginConfig
+	// featureGates is a map of feature names to bools that enable or disable alpha/experimental features.
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/types.go]: Capital F

File: staging/src/k8s.io/kube-scheduler/config/v1alpha1/types.go [hunk 4]
```diff
@@ -104,6 +104,8 @@ type KubeSchedulerConfiguration struct {
 	// PluginConfig is an optional set of custom plugin arguments for each plugin.
 	// Omitting config args for a plugin is equivalent to using the default config for that plugin.
 	PluginConfig []PluginConfig `json:"pluginConfig,omitempty"`
+	// featureGates is a map of feature names to bools that enable or disable alpha/experimental features.
```
Comment (alculquicondor) [inline staging/src/k8s.io/kube-scheduler/config/v1alpha1/types.go]: Capital F

File: cmd/kube-scheduler/app/options/options_test.go [hunk 5]
```diff
@@ -23,15 +23,14 @@ import (
 	"net/http/httptest"
 	"os"
 	"path/filepath"
-	"reflect"
 	"testing"
 	"time"
 
+	"github.com/google/go-cmp/cmp"
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go]: Could you merge the three import blocks into one?

File: cmd/kube-scheduler/app/options/options.go [hunk 6]
```diff
@@ -200,7 +201,7 @@ func (o *Options) ApplyTo(c *schedulerappconfig.Config) error {
 		}
 	}
 
-	return nil
+	return utilfeature.DefaultMutableFeatureGate.SetFromMap(o.ComponentConfig.FeatureGates)
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options.go]: server.go seems like a location more consistent across k8s binaries for this call.

File: cmd/kube-scheduler/app/options/options_test.go [hunk 7]
```diff
@@ -604,7 +619,99 @@ pluginConfig:
...
+			},
+		},
+		{
+			name: "feature gate config",
+			options: &Options{
+				ComponentConfig: func() kubeschedulerconfig.KubeSchedulerConfiguration {
+					cfg, _ := newDefaultComponentConfig()
+					cfg.ClientConnection.Kubeconfig = flagKubeconfig
+					cfg.FeatureGates = map[string]bool{
+						"ResourceLimitsPriorityFunction": true,
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go]: nit: use a different feature gate (from the file test).

File: pkg/scheduler/apis/config/v1alpha1/defaults.go [hunk 8]
```diff
@@ -168,4 +168,7 @@ func SetDefaults_KubeSchedulerConfiguration(obj *kubeschedulerconfigv1alpha1.Kub
 		enableContentionProfiling := true
 		obj.EnableContentionProfiling = &enableContentionProfiling
 	}
+	if obj.FeatureGates == nil {
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1alpha1/defaults.go]: why is this necessary? can we be resilient to nil instead?

**Review by alculquicondor — COMMENTED** [review]
In order to make reviews easier, we ask our contributors to not amend commits so that the context of the review comments is not lost and next reviews are faster.

Once your PR is approved, the approver might ask you to squash.

**Comment by alculquicondor** [comment]
Hi @zouyee, do you have any updates? Thanks

File: pkg/scheduler/apis/config/v1alpha1/defaults.go (hunk 8 above)
Comment (zouyee) [reply pkg/scheduler/apis/config/v1alpha1/defaults.go]: IMO, if we don't test here, we'll have to do the same thing elsewhere

File: cmd/kube-scheduler/app/options/options.go (hunk 6 above)
Comment (zouyee) [reply cmd/kube-scheduler/app/options/options.go]: I'm not so sure about enforcing a section for `FeatureGate` function calls, because it's the last stage of flag processing.

File: cmd/kube-scheduler/app/options/options_test.go (hunk 5 above)
Comment (zouyee) [reply cmd/kube-scheduler/app/options/options_test.go]: the fmt here should be std lib, newline, our stuff, newline, external deps

Comment (alculquicondor) [reply cmd/kube-scheduler/app/options/options_test.go]: uhm... maybe I got confused with the lines that the diff adds. This is correct :)

File: cmd/kube-scheduler/app/options/options.go (hunk 6 above)
Comment (alculquicondor) [reply cmd/kube-scheduler/app/options/options.go]: Sg. I'm don't feel strong about it.

File: pkg/scheduler/apis/config/v1alpha1/defaults.go (hunk 8 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1alpha1/defaults.go]: It seems that this object is only used in `utilfeature.DefaultMutableFeatureGate.SetFromMap`, which uses a `for range`. So it might just work with nil. If it does, we can get rid of all those empty maps in options_test

**Comment by alculquicondor** [comment]
/approve

Please squash

**Comment by alculquicondor** [comment]
/approve cancel

There seems to be an issue of precedence between config file and command line flags. You can still squash

File: cmd/kube-scheduler/app/options/options.go (hunk 6 above)
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options.go]: this should be `c.ComponentConfig`

But there is still problems of precedence that need to be figured out. I'll get back to you in a bit.

Comment (zouyee) [reply cmd/kube-scheduler/app/options/options.go]: Added validation for DefaultMutableFeatureGate

**Comment by alculquicondor** [comment]
This looks good, but I'm waiting for some feedback on #84922

**Comment by zouyee** [comment]
```
ERROR: An error occurred during the fetch of repository 'six':
```
/retest

**Comment by zouyee** [comment]
@alculquicondor PTAL

File: cmd/kube-scheduler/app/options/options_test.go [hunk 9]
```diff
@@ -981,6 +992,99 @@ plugins:
...
+				PodMaxBackoffSeconds:     defaultPodMaxBackoffSeconds,
+				FeatureGates: map[string]bool{
+					"ResourceLimitsPriorityFunction": true,
+				},
+				Profiles: []kubeschedulerconfig.KubeSchedulerProfile{
+					{SchedulerName: "default-scheduler"},
+				},
+			},
+			expectedUsername: "config",
+		},
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go]: what about a test with both (config file and flags)?

File: cmd/kube-scheduler/app/options/options_test.go [hunk 10]
```diff
@@ -737,6 +748,150 @@ profiles:
...
+					QPS:         50,
+					Burst:       100,
+					ContentType: "application/vnd.kubernetes.protobuf",
+				},
+				PercentageOfNodesToScore: defaultPercentageOfNodesToScore,
+				BindTimeoutSeconds:       defaultBindTimeoutSeconds,
+				PodInitialBackoffSeconds: defaultPodInitialBackoffSeconds,
+				PodMaxBackoffSeconds:     defaultPodMaxBackoffSeconds,
+				FeatureGates: map[string]bool{
+					"ResourceLimitsPriorityFunction": true,
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go:887]: This is just taking the values from the file and not the flags.

File: cmd/kube-scheduler/app/options/options.go [hunk 11]
```diff
@@ -195,7 +198,13 @@ func (o *Options) ApplyTo(c *schedulerappconfig.Config) error {
 		if err := validation.ValidateKubeSchedulerConfiguration(cfg).ToAggregate(); err != nil {
 			return err
 		}
-
+		original := o.ComponentConfig.FeatureGates
+		// Add back feature gates that were set in the cfg, but not in flags
+		for k, v := range original {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply cmd/kube-scheduler/app/options/options.go]: I think I was wrong. Don't just take my word, test it :)

[the remark this answers is not in the document]
Comment (alculquicondor) [reply cmd/kube-scheduler/app/options/options.go]: I don't know why I was over complicating things. Flags just override the file. So:

```golang
for k, v := range o.ComponentConfig.FeatureGates {
  cfg.FeatureGates[k] = v
}
```

**Comment by alculquicondor** [comment]
Rebase and make sure suggested tests are in place.

Also, are you moving the feature gates validation to k8s.io/component-base?

**Comment by alculquicondor** [comment]
/close

This is being handled in #92357
